### ISSUE TITLE
Perl crashes reading past the end of a heap block while parsing foreach statement

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7850]
Migrated from [rt.perl.org#34568](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=34568) (status was 'resolved')

Searchable as RT34568$

**Comment by p5pRT** [comment]
### From John_Sargent@McAfee.com

#### Created by john_sargent@mcafee.com



In op\.c\, line 3856 \(ish\) the cast to a LOOP\* is incorrect\. in that the
memory allocated inside Perl\_convert when the OP type is NOT OP\_LIST is 
less than the size of a LOOP structure\. Consequently\, if the Copy call
at line 3846 \(ish\) it will try to read beyond the memory allocated\. 

If this memory happens to be the first element in the allocated chunk\,
it will be reading memory that has not been allocated at all\. So\, at 
best it reads garbage\, or as happens several times here\, it crashes
with an access violation\.


OP\* Perl\_newFOROP\(\.\.\.\)
\{
&nbsp;  LOOP  \*loop;

&nbsp;    \<\-snip\->

 \<\<\<\< MEMORY ALLOCATED IN CONVERT HERE IS NOT BIG ENOUGH FOR A LOOP
STRUCT

&nbsp;  loop = \(LOOP\*\)list\(convert\(OP\_ENTERITER\, iterflags\,   
&nbsp;                 append\_elem\(OP\_LIST\, expr\, scalar\(sv\)\)\)\);
&nbsp;  assert\(\!loop\->op\_next\);

&nbsp;    \<\-snip\->


&nbsp;  LOOP \*tmp;
&nbsp;  NewOp\(1234\,tmp\,1\,LOOP\);
&nbsp;  Copy\(loop\,tmp\,1\,LOOP\);     \<\<\<\<\< ACCESS VIOLATION
&nbsp;  FreeOp\(loop\);
&nbsp;  loop= tmp;


&nbsp;       \<\-snip\->

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=critical

Site configuration information for perl v5.8.0:

Configured by bhcompile at Tue Feb 18 22:17:47 EST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.20-2.48smp,
archname=i386-linux-thread-multi
    uname='linux stripples.devel.redhat.com 2.4.20-2.48smp #1 smp thu
feb 13 11:44:55 est 2003 i686 i686 i386 gnulinux '
    config_args='-des -Doptimize=-O2 -march=i386 -mcpu=i686 -g
-Dmyhostname=localhost -Dperladmin=root@localhost -Dcc=gcc -Dcf_by=Red
Hat, Inc. -Dinstallprefix=/usr -Dprefix=/usr -Darchname=i386-linux
-Dvendorprefix=/usr -Dsiteprefix=/usr
-Dotherlibdirs=/usr/lib/perl5/5.8.0 -Duseshrplib -Dusethreads
-Duseithreads -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Di_db
-Ui_ndbm -Di_gdbm -Di_shadow -Di_syslog -Dman3ext=3pm -Duseperlio
-Dinstallusrbinperl -Ubincompat5005 -Uversiononly -Dpager=/usr/bin/less
-isr'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS
-DDEBUGGING -fno-strict-aliasing -I/usr/local/include
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2 -march=i386 -mcpu=i686 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBUGGING
-fno-strict-aliasing -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='3.2.2 20030213 (Red Hat Linux 8.0
3.2.2-1)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.3.1.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.3.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic
-Wl,-rpath,/usr/lib/perl5/5.8.0/i386-linux-thread-multi/CORE'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    MAINT18379


@INC for perl v5.8.0:
    /usr/lib/perl5/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.0
    /usr/lib/perl5/vendor_perl
    /usr/lib/perl5/5.8.0/i386-linux-thread-multi
    /usr/lib/perl5/5.8.0
    .


Environment for perl v5.8.0:
    HOME=/home/jsargent
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/home/jsargent/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Sargent\, John wrote&#8203;:
> In op\.c\, line 3856 \(ish\) the cast to a LOOP\* is incorrect\. in that the
> memory allocated inside Perl\_convert when the OP type is NOT OP\_LIST is 

I can't figure out a case where this wouldn't be a list op\.
Have you code to reproduce this ?

> less than the size of a LOOP structure\. Consequently\, if the Copy call
> at line 3846 \(ish\) it will try to read beyond the memory allocated\. 
> 
> If this memory happens to be the first element in the allocated chunk\,
> it will be reading memory that has not been allocated at all\. So\, at 
> best it reads garbage\, or as happens several times here\, it crashes
> with an access violation\.
> 
> 
> OP\* Perl\_newFOROP\(\.\.\.\)
> \{
>   LOOP  \*loop;
> 
>     \<\-snip\->
> 
>  \<\<\<\< MEMORY ALLOCATED IN CONVERT HERE IS NOT BIG ENOUGH FOR A LOOP
> STRUCT
> 
>   loop = \(LOOP\*\)list\(convert\(OP\_ENTERITER\, iterflags\,   
>                  append\_elem\(OP\_LIST\, expr\, scalar\(sv\)\)\)\);

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From John_Sargent@McAfee.com

If scalar\(sv\) is null \(not sure why it would be off hand\)\, and expr is
not 
a list type\, append will just return the expr object\, which is not of
type
OP\_LIST\. Perl\_convert will then return a new LISTOP\. 

The crash is because the code casts a LISTOP to a LOOP which is bigger
as Copy
takes the expected type as an argument\, and assumes the data size from
that
rather than the actual size of the data\.



\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Rafael Garcia\-Suarez via RT \[mailto&#8203;:perlbug\-followup@&#8203;perl\.org\] 
Sent&#8203;: 16 March 2005 12&#8203;:02
To&#8203;: Sargent\, John
Subject&#8203;: Re&#8203;: \[perl \#34450\] Coding defect in Perl\_newFOROP causing Access
Violation


Sargent\, John wrote&#8203;:
> In op\.c\, line 3856 \(ish\) the cast to a LOOP\* is incorrect\. in that the
> memory allocated inside Perl\_convert when the OP type is NOT OP\_LIST
is 

I can't figure out a case where this wouldn't be a list op\.
Have you code to reproduce this ?

> less than the size of a LOOP structure\. Consequently\, if the Copy call
> at line 3846 \(ish\) it will try to read beyond the memory allocated\. 
> 
> If this memory happens to be the first element in the allocated chunk\,
> it will be reading memory that has not been allocated at all\. So\, at 
> best it reads garbage\, or as happens several times here\, it crashes
> with an access violation\.
> 
> 
> OP\* Perl\_newFOROP\(\.\.\.\)
> \{
>   LOOP  \*loop;
> 
>     \<\-snip\->
> 
>  \<\<\<\< MEMORY ALLOCATED IN CONVERT HERE IS NOT BIG ENOUGH FOR A LOOP
> STRUCT
> 
>   loop = \(LOOP\*\)list\(convert\(OP\_ENTERITER\, iterflags\,   
>                  append\_elem\(OP\_LIST\, expr\, scalar\(sv\)\)\)\);

**Comment by p5pRT** [comment]
### From spenlow@exchange.microsoft.com

This is a bug report for perl from spenlow@&#8203;microsoft\.com\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.3\.




---
With version 809 of ActiveState Perl for Win32 I am
intermittently encountering a crash at this callstack \(I can
also repro this with build 808\)&#8203;:

MSVCRT\!memcpy\+0x33
perl58\!Perl\_newFOROP\+0x1da \[\.\.\\op\.c @&#8203; 3846\]
perl58\!Perl\_yyparse\+0x4d6 \[perly\.y @&#8203; 264\]
perl58\!S\_parse\_body\+0x9e7 \[\.\.\\perl\.c @&#8203; 1701\]
perl58\!perl\_parse\+0x289 \[\.\.\\perl\.c @&#8203; 1213\]
perl58\!RunPerl\+0x77 \[perllib\.c @&#8203; 195\]
perl\!main\+0x12 \[perlmain\.c @&#8203; 17\]

Perl appears to be parsing a foreach statement\. While it's doing
that it copies from one data structure to another and it reads
past the end of the block\. The block is allocated here&#8203;:

MSVCRT\!malloc\+0x74
perl58\!VMem&#8203;::Malloc\+0x14 \[vmem\.h @&#8203; 164\]
perl58\!CPerlHost&#8203;::MallocShared\+0x1d \[perlhost\.h @&#8203; 82\]
perl58\!PerlMemSharedMalloc\+0x10 \[perlhost\.h @&#8203; 346\]
perl58\!Perl\_Slab\_Alloc\+0x28 \[\.\.\\op\.c @&#8203; 44\]
perl58\!Perl\_newLISTOP\+0x15 \[\.\.\\op\.c @&#8203; 2204\]
perl58\!Perl\_append\_elem\+0x5b \[\.\.\\op\.c @&#8203; 2112\]
perl58\!Perl\_newFOROP\+0x1a4 \[\.\.\\op\.c @&#8203; 3837\]
perl58\!Perl\_yyparse\+0x4d6 \[perly\.y @&#8203; 264\]
perl58\!S\_parse\_body\+0x9e7 \[\.\.\\perl\.c @&#8203; 1701\]
perl58\!perl\_parse\+0x289 \[\.\.\\perl\.c @&#8203; 1213\]
perl58\!RunPerl\+0x77 \[perllib\.c @&#8203; 195\]
perl\!main\+0x12 \[perlmain\.c @&#8203; 17\]

The block appears to be 0x200C in length\.

This problem occurs intermittently on many different machines\.
I was able to get a 100% repro situation by creating a lot of
environment variables\. My hunch is that this causes more
memory to be used in the heap\, such that this allocation of
0x200C is aligned at the end of the page\. Then when perl reads
past the end\, it crashes\.

My hunch of the problem is that Perl\_Slab\_Alloc\, VMem&#8203;::Malloc\,
CPerlHost&#8203;::MallocShared\, or PerlMemSharedMalloc is doing some
math to round\-up/down or add a header/trailer and the math
isn't always correct\.

This bug is very bad because it is intermittent and hard to
repro\. i\.e\. I have a machine in my office with a script that
can repro this\, but at the moment I can't distribute the script
and even if I did\, you would probably have problems reproing
the issue \(i\.e\. I had to work for a few hours to get my machine
in a state where I could repro the issue\)\.

I would be happy to talk to anyone about this issue or answer
any questions\.

Thank you\,

Spencer Low
spencer@&#8203;microsoft\.com
JAPH ;\-\)



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=high


---
Site configuration information for perl v5\.8\.3&#8203;:

Configured by ActiveState at Tue Feb  3 00&#8203;:28&#8203;:38 2004\.

Summary of my perl5 \(revision 5 version 8 subversion 3\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=MSWin32\, osvers=4\.0\, archname=MSWin32\-x86\-multi\-thread
&nbsp;    uname=''
&nbsp;    config\_args='undef'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=undef
&nbsp;    usethreads=undef use5005threads=undef useithreads=define
usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cl'\, ccflags ='\-nologo \-Gf \-W3 \-MD \-Zi \-DNDEBUG \-O1 \-DWIN32
\-D\_CONSOLE \-DNO\_STRICT \-DHAVE\_DES\_FCRYPT  \-DNO\_HASH\_SEED
\-DPERL\_IMPLICIT\_CONTEXT \-DPERL\_IMPLICIT\_SYS \-DUSE\_PERLIO
\-DPERL\_MSVCRT\_READFIX'\,
&nbsp;    optimize='\-MD \-Zi \-DNDEBUG \-O1'\,
&nbsp;    cppflags='\-DWIN32'
&nbsp;    ccversion=''\, gccversion=''\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=undef\, longlongsize=8\, d\_longdbl=define\, longdblsize=10
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='\_\_int64'\,
lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='link'\, ldflags ='\-nologo \-nodefaultlib \-debug \-opt&#8203;:ref\,icf
\-libpath&#8203;:"C&#8203;:\\Perl\\lib\\CORE"  \-machine&#8203;:x86'
&nbsp;    libpth=C&#8203;:\\PROGRA~1\\MICROS~3\\VC98\\lib
&nbsp;    libs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib winspool\.lib
comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib oleaut32\.lib
netapi32\.lib uuid\.lib wsock32\.lib mpr\.lib winmm\.lib  version\.lib
odbc32\.lib odbccp32\.lib msvcrt\.lib
&nbsp;    perllibs=  oldnames\.lib kernel32\.lib user32\.lib gdi32\.lib
winspool\.lib  comdlg32\.lib advapi32\.lib shell32\.lib ole32\.lib
oleaut32\.lib  netapi32\.lib uuid\.lib wsock32\.lib mpr\.lib winmm\.lib
version\.lib odbc32\.lib odbccp32\.lib msvcrt\.lib
&nbsp;    libc=msvcrt\.lib\, so=dll\, useshrplib=yes\, libperl=perl58\.lib
&nbsp;    gnulibc\_version='undef'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_win32\.xs\, dlext=dll\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags='\-dll \-nologo \-nodefaultlib \-debug
\-opt&#8203;:ref\,icf  \-libpath&#8203;:"C&#8203;:\\Perl\\lib\\CORE"  \-machine&#8203;:x86'

Locally applied patches&#8203;:
&nbsp;    ACTIVEPERL\_LOCAL\_PATCHES\_ENTRY
&nbsp;    22218 Remove the caveat about detached threads crashing on Windows
&nbsp;    22201 Avoid threads\+win32 crash by freeing Perl interpreter slightly
later
&nbsp;    22169 Display 'out of memeory' errors using low\-level I/O
&nbsp;    22159 Upgrade to Time&#8203;::Hires 1\.55
&nbsp;    22120 Make 'Configure \-Dcf\_by=\.\.\.' work
&nbsp;    22051 Upgrade to Time&#8203;::HiRes 1\.54
&nbsp;    21540 Fix backward\-compatibility issues in if\.pm



---
@&#8203;INC for perl v5\.8\.3&#8203;:
&nbsp;    E&#8203;:\\sqlmain\\tools\\winsdk\\tools\\perllib
&nbsp;    E&#8203;:\\sqlmain\\tools\\winsdk\\tools\\perl\\site\\lib
&nbsp;    E&#8203;:\\sqlmain\\tools\\winsdk\\tools\\perl\\lib
&nbsp;    E&#8203;:/sqlmain/tools/winsdk/tools/perl/lib
&nbsp;    E&#8203;:/sqlmain/tools/winsdk/tools/perl/site/lib
&nbsp;    \.



---
Environment for perl v5\.8\.3&#8203;:
&nbsp;    HOME \(unset\)
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
 
PATH=E&#8203;:\\sqlmain\\tools\\winsdk\\tools\\perl\\bin;C&#8203;:\\WINDOWS\\system32;C&#8203;:\\WINDO
WS;C&#8203;:\\WINDOWS\\system32\\WBEM;E&#8203;:\\tools;C&#8203;:\\Program Files\\Support
Tools\\;D&#8203;:\\Debuggers;
 
PERL5LIB=E&#8203;:\\sqlmain\\tools\\winsdk\\tools\\perllib;E&#8203;:\\sqlmain\\tools\\winsdk\\t
ools\\perl\\site\\lib;E&#8203;:\\sqlmain\\tools\\winsdk\\tools\\perl\\lib
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL \(unset\)

**Comment by p5pRT** [comment]
### From @gsar

On 25 Mar 2005 07&#8203;:54&#8203;:22 GMT\, "Spencer Low" wrote&#8203;:
>\# New Ticket Created by  "Spencer Low" 
>\# Please include the string&#8203;:  \[perl \#34568\]
>\# in the subject line of all future correspondence about this issue\. 
>\# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=34568 >

This appears to have been previously reported as bug 34450&#8203;:

&nbsp;  http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2005\-03/msg00339\.html
&nbsp;  https://rt-archive.perl.org/perl5/Ticket/Display.html?id=34450

Some more history&#8203;: the buggy code has been around for a long time\,
but was only enabled if PL\_OP\_SLAB\_ALLOC was \#defined\.  It got
enabled by default for PERL\_IMPLICIT\_SYS builds in change 14284\,
which shipped beginning with 5\.8\.0\.

>With version 809 of ActiveState Perl for Win32 I am
>intermittently encountering a crash at this callstack \(I can
>also repro this with build 808\)&#8203;:
>
>MSVCRT\!memcpy\+0x33
>perl58\!Perl\_newFOROP\+0x1da \[\.\.\\op\.c @&#8203; 3846\]
>perl58\!Perl\_yyparse\+0x4d6 \[perly\.y @&#8203; 264\]
>perl58\!S\_parse\_body\+0x9e7 \[\.\.\\perl\.c @&#8203; 1701\]
>perl58\!perl\_parse\+0x289 \[\.\.\\perl\.c @&#8203; 1213\]
>perl58\!RunPerl\+0x77 \[perllib\.c @&#8203; 195\]
>perl\!main\+0x12 \[perlmain\.c @&#8203; 17\]

Nice analysis\.  The attached patch against perl 5\.8\.6 might fix it\.


Sarathy
gsar@&#8203;ActiveState\.com
<details><summary>Inline Patch</summary>

```diff
-----------------------------------8<-----------------------------------
Index: ./op.c
--- ./op.c.~1~	Fri Mar 25 10:25:09 2005
+++ ./op.c	Fri Mar 25 10:25:09 2005
@@ -3930,7 +3930,7 @@
     {
 	LOOP *tmp;
 	NewOp(1234,tmp,1,LOOP);
-	Copy(loop,tmp,1,LOOP);
+	Copy(loop,tmp,1,LISTOP);
 	FreeOp(loop);
 	loop = tmp;
     }
End of Patch.
```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rgarcia

On Fri\, 25 Mar 2005 10&#8203;:31&#8203;:09 \-0800\, Gurusamy Sarathy
\<gsar@&#8203;activestate\.com> wrote&#8203;:
> This appears to have been previously reported as bug 34450&#8203;:
> 
>   http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2005\-03/msg00339\.html
>   https://rt-archive.perl.org/perl5/Ticket/Display.html?id=34450

But\, as far as I can tell\, in but 34450 PL\_OP\_SLAB\_ALLOC isn't defined\.\.\.

> Some more history&#8203;: the buggy code has been around for a long time\,
> but was only enabled if PL\_OP\_SLAB\_ALLOC was \#defined\.  It got
> enabled by default for PERL\_IMPLICIT\_SYS builds in change 14284\,
> which shipped beginning with 5\.8\.0\.
> 
> >With version 809 of ActiveState Perl for Win32 I am
> >intermittently encountering a crash at this callstack \(I can
> >also repro this with build 808\)&#8203;:
> >
> >MSVCRT\!memcpy\+0x33
> >perl58\!Perl\_newFOROP\+0x1da \[\.\.\\op\.c @&#8203; 3846\]
> >perl58\!Perl\_yyparse\+0x4d6 \[perly\.y @&#8203; 264\]
> >perl58\!S\_parse\_body\+0x9e7 \[\.\.\\perl\.c @&#8203; 1701\]
> >perl58\!perl\_parse\+0x289 \[\.\.\\perl\.c @&#8203; 1213\]
> >perl58\!RunPerl\+0x77 \[perllib\.c @&#8203; 195\]
> >perl\!main\+0x12 \[perlmain\.c @&#8203; 17\]
> 
> Nice analysis\.  The attached patch against perl 5\.8\.6 might fix it\.
> 
> Sarathy
> gsar@&#8203;ActiveState\.com
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-8\<\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> Index&#8203;: \./op\.c
> \-\-\- \./op\.c\.~1~  Fri Mar 25 10&#8203;:25&#8203;:09 2005
> \+\+\+ \./op\.c      Fri Mar 25 10&#8203;:25&#8203;:09 2005
> @&#8203;@&#8203; \-3930\,7 \+3930\,7 @&#8203;@&#8203;
>      \{
>         LOOP \*tmp;
>         NewOp\(1234\,tmp\,1\,LOOP\);
> \-       Copy\(loop\,tmp\,1\,LOOP\);
> \+       Copy\(loop\,tmp\,1\,LISTOP\);
>         FreeOp\(loop\);
>         loop = tmp;
>      \}
> End of Patch\.
>

**Comment by p5pRT** [comment]
### From @gsar

On Fri\, 25 Mar 2005 20&#8203;:25&#8203;:35 \+0100\, Rafael Garcia\-Suarez wrote&#8203;:
>On Fri\, 25 Mar 2005 10&#8203;:31&#8203;:09 \-0800\, Gurusamy Sarathy
>\<gsar@&#8203;activestate\.com> wrote&#8203;:
>> This appears to have been previously reported as bug 34450&#8203;:
>> 
>>   http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2005\-03/msg00339\.ht
>>   https://rt-archive.perl.org/perl5/Ticket/Display.html?id=34450
>
>But\, as far as I can tell\, in but 34450 PL\_OP\_SLAB\_ALLOC isn't defined\.\.\.

I don't think that is possible\, since the code in question is
protected by an \#ifdef PL\_OP\_SLAB\_ALLOC\.


Sarathy
gsar@&#8203;ActiveState\.com

**Comment by p5pRT** [comment]
### From @rgarcia

On Fri\, 25 Mar 2005 12&#8203;:54&#8203;:39 \-0800\, Gurusamy Sarathy
\<gsar@&#8203;activestate\.com> wrote&#8203;:
> >But\, as far as I can tell\, in but 34450 PL\_OP\_SLAB\_ALLOC isn't defined\.\.\.
> 
> I don't think that is possible\, since the code in question is
> protected by an \#ifdef PL\_OP\_SLAB\_ALLOC\.

Yes\, that's more or less what I was saying\, thus wondering about the
link with 34450\.
Anyway\, thanks\, applied to blead as \#24080\.

**Comment by p5pRT** [comment]
### From @smpeters

> \[rgarciasuarez@&#8203;gmail\.com \- Fri Mar 25 13&#8203;:37&#8203;:04 2005\]&#8203;:
> 
> On Fri\, 25 Mar 2005 12&#8203;:54&#8203;:39 \-0800\, Gurusamy Sarathy
> \<gsar@&#8203;activestate\.com> wrote&#8203;:
> > >But\, as far as I can tell\, in but 34450 PL\_OP\_SLAB\_ALLOC isn't
defined\.\.\.
> > 
> > I don't think that is possible\, since the code in question is
> > protected by an \#ifdef PL\_OP\_SLAB\_ALLOC\.
> 
> Yes\, that's more or less what I was saying\, thus wondering about the
> link with 34450\.
> Anyway\, thanks\, applied to blead as \#24080\.
> 
> 
Ticket closed as patch was applied\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From spenlow@exchange.microsoft.com

Thank you for the fast reply and patch\. When is the next release of
ActivePerl going to include this fix? aka\, how can I get binaries with
this fix without building perl myself?

Thanks again for the fast response\,

Spencer

\-\-\-\-\-Original Message\-\-\-\-\-
From&#8203;: Gurusamy Sarathy \[mailto&#8203;:gsar@&#8203;ActiveState\.com\] 
Sent&#8203;: Friday\, March 25\, 2005 10&#8203;:31 AM
To&#8203;: perl5\-porters@&#8203;perl\.org
Cc&#8203;: bugs\-bitbucket@&#8203;netlabs\.develooper\.com; john\_sargent@&#8203;mcafee\.com;
Spencer Low
Subject&#8203;: Re&#8203;: \[perl \#34568\] Perl crashes reading past the end of a heap
block while parsing foreach statement 

On 25 Mar 2005 07&#8203;:54&#8203;:22 GMT\, "Spencer Low" wrote&#8203;:
>\# New Ticket Created by  "Spencer Low" 
>\# Please include the string&#8203;:  \[perl \#34568\] \# in the subject line of 
>all future correspondence about this issue\.
>\# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=34568 >

This appears to have been previously reported as bug 34450&#8203;:

 
http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2005\-03/msg00339\.
html
&nbsp;  https://rt-archive.perl.org/perl5/Ticket/Display.html?id=34450

Some more history&#8203;: the buggy code has been around for a long time\, but
was only enabled if PL\_OP\_SLAB\_ALLOC was \#defined\.  It got enabled by
default for PERL\_IMPLICIT\_SYS builds in change 14284\, which shipped
beginning with 5\.8\.0\.

>With version 809 of ActiveState Perl for Win32 I am intermittently 
>encountering a crash at this callstack \(I can also repro this with 
>build 808\)&#8203;:
>
>MSVCRT\!memcpy\+0x33
>perl58\!Perl\_newFOROP\+0x1da \[\.\.\\op\.c @&#8203; 3846\]
>perl58\!Perl\_yyparse\+0x4d6 \[perly\.y @&#8203; 264\]
>perl58\!S\_parse\_body\+0x9e7 \[\.\.\\perl\.c @&#8203; 1701\]
>perl58\!perl\_parse\+0x289 \[\.\.\\perl\.c @&#8203; 1213\]
>perl58\!RunPerl\+0x77 \[perllib\.c @&#8203; 195\]
>perl\!main\+0x12 \[perlmain\.c @&#8203; 17\]

Nice analysis\.  The attached patch against perl 5\.8\.6 might fix it\.


Sarathy
gsar@&#8203;ActiveState\.com
<details><summary>Inline Patch</summary>

```diff
-----------------------------------8<-----------------------------------
Index: ./op.c
--- ./op.c.~1~	Fri Mar 25 10:25:09 2005
+++ ./op.c	Fri Mar 25 10:25:09 2005
@@ -3930,7 +3930,7 @@
     {
 	LOOP *tmp;
 	NewOp(1234,tmp,1,LOOP);
-	Copy(loop,tmp,1,LOOP);
+	Copy(loop,tmp,1,LISTOP);
 	FreeOp(loop);
 	loop = tmp;
     }
End of Patch.
```
</details>

**Comment by p5pRT** [comment]
### From @smpeters

> \[John\_Sargent@&#8203;McAfee\.com \- Wed Mar 16 06&#8203;:02&#8203;:13 2005\]&#8203;:
> 
> If scalar\(sv\) is null \(not sure why it would be off hand\)\, and expr is
> not 
> a list type\, append will just return the expr object\, which is not of
> type
> OP\_LIST\. Perl\_convert will then return a new LISTOP\. 
> 
> The crash is because the code casts a LISTOP to a LOOP which is bigger
> as Copy
> takes the expected type as an argument\, and assumes the data size from
> that
> rather than the actual size of the data\.
> 
> 
> 
> \-\-\-\-\-Original Message\-\-\-\-\-
> From&#8203;: Rafael Garcia\-Suarez via RT \[mailto&#8203;:perlbug\-followup@&#8203;perl\.org\] 
> Sent&#8203;: 16 March 2005 12&#8203;:02
> To&#8203;: Sargent\, John
> Subject&#8203;: Re&#8203;: \[perl \#34450\] Coding defect in Perl\_newFOROP causing Access
> Violation
> 
> 
> Sargent\, John wrote&#8203;:
> > In op\.c\, line 3856 \(ish\) the cast to a LOOP\* is incorrect\. in that the
> > memory allocated inside Perl\_convert when the OP type is NOT OP\_LIST
> is 
> 
> I can't figure out a case where this wouldn't be a list op\.
> Have you code to reproduce this ?
> 

Let me clarify\.  Do you have any Perl and/or XS code that demonstrates this problem?

**Comment by p5pRT** [comment]
### From @smpeters

> \[stmpeters \- Thu Jul 14 20&#8203;:39&#8203;:33 2005\]&#8203;:
> 
> > \[John\_Sargent@&#8203;McAfee\.com \- Wed Mar 16 06&#8203;:02&#8203;:13 2005\]&#8203;:
> >
> > If scalar\(sv\) is null \(not sure why it would be off hand\)\, and expr
> is
> > not
> > a list type\, append will just return the expr object\, which is not
> of
> > type
> > OP\_LIST\. Perl\_convert will then return a new LISTOP\.
> >
> > The crash is because the code casts a LISTOP to a LOOP which is
> bigger
> > as Copy
> > takes the expected type as an argument\, and assumes the data size
> from
> > that
> > rather than the actual size of the data\.
> >
> >
> >
> > \-\-\-\-\-Original Message\-\-\-\-\-
> > From&#8203;: Rafael Garcia\-Suarez via RT \[mailto&#8203;:perlbug\-followup@&#8203;perl\.org\]
> > Sent&#8203;: 16 March 2005 12&#8203;:02
> > To&#8203;: Sargent\, John
> > Subject&#8203;: Re&#8203;: \[perl \#34450\] Coding defect in Perl\_newFOROP causing
> Access
> > Violation
> >
> >
> > Sargent\, John wrote&#8203;:
> > > In op\.c\, line 3856 \(ish\) the cast to a LOOP\* is incorrect\. in that
> the
> > > memory allocated inside Perl\_convert when the OP type is NOT
> OP\_LIST
> > is
> >
> > I can't figure out a case where this wouldn't be a list op\.
> > Have you code to reproduce this ?
> >
> 
> Let me clarify\.  Do you have any Perl and/or XS code that demonstrates
> this problem?
> 
Without a demonstration of the problem mentioned\, this ticket will be
set to stalled\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'stalled'

**Comment by p5pRT** [comment]
### From @doy

If no code can be provided to demonstrate this issue on a more recent
version of perl\, this ticket can probably be closed\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'stalled' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jun 24 15&#8203;:21&#8203;:47 2012\, doy wrote&#8203;:
> If no code can be provided to demonstrate this issue on a more recent
> version of perl\, this ticket can probably be closed\.

It was fixed in commit bd5f3bc4\, so this is the same bug as \#34568\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
